source/GMController.cpp

```diff
@@ -1 +1,68 @@
-#include "GMController.h"
+﻿#include "GMController.h"
+
+using namespace cugl;
+using namespace std;
+
+#pragma mark -
+#pragma mark GM Factors
+
+/** The maximum number of events on ship at any one time. This will probably need to scale with the
+ * number of players*/
+constexpr unsigned int MAX_EVENTS = 3;
+
+#pragma mark -
+#pragma mark GM
+/**
+ * Creates a new GM controller.
+ *
+ * This constructor does NOT do any initialzation.  It simply allocates the
+ * object. This makes it safe to use this class without a pointer.
+ */
+GMController::GMController() : active(false), numEvents(0) {}
+
+/**
+ * Deactivates this input controller, releasing all listeners.
+ *
+ * This method will not dispose of the input controller. It can be reused
+ * once it is reinitialized.
+ */
+void GMController::dispose() {
+	if (active) {
+		active = false;
+	}
+}
+
+/**
+ * Initializes the GM
+ *
+ * This method works like a proper constructor, initializing the GM
+ * controller and allocating memory.
+ *
+ * @return true if the controller was initialized successfully
+ */
+bool GMController::init(const std::vector<std::shared_ptr<BreachModel>> b) {
+	bool success = true;
+	breaches = b;
+	// Set random seed based on time
+	srand(time(NULL));
+	active = success;
+	return success;
+}
+
+/**
+ * Processes the GM.
+ *
+ * This method is used to run the GM for generating and managing current ship events
+ */
+void GMController::update(float dt) {
+	// Simple logic for adding breaches when under max, replace with actual logic later
+	if (numEvents < MAX_EVENTS) {
+		breaches.at(numEvents)->setAngle((rand() % 360) * (float)M_PI / 180.0f);
+		numEvents++;
+	}
+}
```

Comment:
Do we want some randomization on when to add events, so that we're not permanently locked in at MAX_EVENTS?

Reply:
I put it in for now with a constant, but eventually the number should just automatically scale with the number of players and placement of breaches shouldn't just spam them until at max randomly. I figured for now this was good enough for gameplay prototype for spawning. We might want to figure out how tap is registered in the scenegraph tree and just let it spawn like this.

Reply:
I mean that the update method should maybe roll a random number to determine whether or not to add an event even if we're under MAX_EVENTS; otherwise, we'd be permanently at the max all the time. It should only take like a single line if statement; `if(rand() < ???) return;` maybe. Just an idea. I'll let the other reviewers weigh in too.

Reply:
Oh ok, yeah we can just put like a 5% chance or something on every frame.

Reply:
hmm, just checked a binomical calculator, and at 60 update loops a second and a 5% chance it would be a ~95% chance to get a breach. Gonna set it to .5% chance for a ~26% chance to get a breach every second.

Reply:
But also since we aren't removing any, it'll still only take a few seconds to cap out at 3 breaches

Reply:
I think a few seconds is a fine timeframe for a small ship like we currently have.

Reply:
Turns out 1/1000 only takes like 20 seconds to spawn all 3 breaches. So I went with that since thats about 3 rounds around the ship